### ISSUE TITLE
can't reproduce results after restore a model

### ISSUE THREAD

**Issue by davide1993** [autonomio/talos#187]
After a scan I save best models for metric 'val_acc', 'val_loss' ... 
When I try to restore a model and pick the best for 'val_acc' metric, if I test that model on train data and validation data I get different results than report why?

```
ta.Deploy(scan,exp_acc_filename,metric= val_acc)
r_acc4=ta.Restore(path)
model4=r_acc4.model
y_pred_train=model4.predict_classes(X_train)
accuracy_score(y_train,y_pred_train) #output 90% but should be 92%
```

**Comment by mikkokotila** [comment]
Can you try with the latest daily-dev version. A fix was made yesterday that might have to do with this: 

```
pip uninstall talos
pip install git+https://github.com/autonomio/talos@daily-dev
```

Remember to uninstall first, as the version number have not changed. What do you get then?

**Comment by davide1993** [comment]
sorry, but what about all my tests? Now I try to reinstall talos

**Comment by mikkokotila** [comment]
What do you mean "what about all my tests"? 

Replacing current version with a newer version is a standard procedure. It is backward compatible.

**Comment by davide1993** [comment]
Nothing changed from before.
I did a lot of test and after week of work If results are not reproducible, for my thesis is really bad.

**Comment by davide1993** [comment]
maybe i'm using not the same function talos use?

**Comment by mikkokotila** [comment]
In order for the effect of the change to take place, you have to run best_model() in the newly updated version, and then use that model for predictions.

Also there is an important notion regarding "reproducibility" with deep learning models. Basically you're able to reproduce results only when the model is not changed (no retraining) and you apply exactly the same dataset to it. So if the model is the same, and the dataset is the same, you get exactly the same result. Otherwise you will not get the same result. Just because you train the model with exactly the same hyperparameters does not mean you end up with exactly the same result because there is an element of randomness involved in the process. Maybe if you explain exactly what you are doing and trying to achieve, I might be able to help.

**Comment by mikkokotila** [comment]
re: maybe i'm using not the same function talos use?

...what do you mean by this?

**Comment by davide1993** [comment]
in csv file created after the experiment there are some columns like val_acc val_loss ecc... 
If I restore the model and make prediction with that model on the same train and validation set using predict_classes and then accuracy_score(y_train,y_pred_train) I get different results. Maybe the scan procedure don't use the same logic??? The problem is that my restored model don't output the results obtained in the experiment. What can I do?

**Comment by mikkokotila** [comment]
When you conduct the experiment Scan() you are giving it some data, let it be called 'dataset-A' and then at the end of the experiment, you have the results of all the permutations in the experiment. Each of these constitute a Keras model with a given configuration. Within dataset-A you have two results; training results (e.g. acc and loss) and validation results (e.g. val_acc and val_loss). 

Then when you access the `scan_object.best_model(metric='val_acc')` it goes to the results of the experiment (as per above) and takes saved model based on 'val_acc' and activates it so you can use it for making predictions. 

So now you have a model, and you know how that model have performed in training and validation with dataset-A. If you now input some other dataset to it, say dataset-B you will not get the same result, nor should you get it. That's the first part of this, just to make sure it's clear. 

The second part relates with `accuracy_score`. If this is from sklearn, then it is a known issue that it will not give the same result for you as the internal keras accuracy. In that case maybe [this issue](https://github.com/keras-team/keras/issues/9672) is relevant. But you can find many more in Stackoverflow and elsewhere. 

Also, don't worry, you'll end up hitting a lot of snags as a researcher ... we are after all operating in a vanguard area of numerical computing ... but generally it's straightforward to overcome whatever issues may face in this field :)

**Comment by davide1993** [comment]
I use the same train, validation and test in all the experiments, this is sure. I don't use scan object to take the best model because I need to pick up the best model after days or when needed (not run-time), so I save best model with deploy. If I restore the model from the zip file created with deploy, I want to reproduce the same results that model show before, but I can't what can I do? My problem is a binary classification

**Comment by mikkokotila** [comment]
Is the `accuracy_score` from sklearn?

**Comment by mikkokotila** [comment]
Also maybe you can try `evaluate()` from Keras  instead of `accuracy_score()` from Sklearn (I'm assuming because of your issue and the function name that you are actually using Sklearn).

**Comment by davide1993** [comment]
1) I use sklearn.metrics.accuracy_score
2) all datasets are identical, I can't also reproduce train set results not only on validation...

**Comment by mikkokotila** [comment]
The case you are reporting is a known issue related with a discrepancy between Keras and Sklearn way to do binary accuracy. It does not have anything to do with Talos. You can use Keras evaluate() instead of Sklearn accuracy_score(). 

You can use Google to find many information about this topic. You could start [here](https://www.google.com/search?q=sklearn+accuracy_score+different+result+with+Keras&oq=sklearn+accuracy_score+different+result+with+Keras&aqs=chrome..69i57.9817j1j7&sourceid=chrome&ie=UTF-8). 

Closing here. Feel free to open a new issue if anything else comes up.

**Comment by davide1993** [comment]
I test now keras evaluate, it prints exactly the same results as Sklearn accuracy_score()

**Comment by davide1993** [comment]
I try to run scan with the data but now I get 70% val_accuracy instead 92%, I don't know what to do

**Comment by mikkokotila** [comment]
Are you using a notebook? If yes, can you share the notebook you are using to run your experiment. Or otherwise create a [dump](http://pastebin.com) with the required codes, and I will have a look.

**Comment by davide1993** [comment]
I'll give you the train, val, and test set and the zip generated from deploy, do you need other?
thank you

**Comment by davide1993** [comment]
Ok I need some minutes, i'll give you everything as soon as possible

**Comment by davide1993** [comment]
https://github.com/davide1993/talosExperiment
you can find everything here

**Comment by davide1993** [comment]
I reupload the notebook because there was a wrong comment, now you can pull

**Comment by mikkokotila** [comment]
Thanks. I've looked into it now. 

The way scan_object.data and the accompanying .csv gets the results from metrics is through the history object that comes out from model.fit() in Keras. There are two modes in which Talos can do this: 

- based on highest epoch (which is default)
- based on last epoch(you have to set` Scan( ... last_epoch_value=True ... )`)

As I have seen, in your model you are currently using `last_epoch_value=False` which means you are unlikely to ever get exactly the same result from evaluation as you will get from the Scan() results. I think based on this experience, we will have it True by default going forward. Still then you will end up with the problem, which has to do with Keras, where whatever you have as your last epoch metrics, is not going to be the same as the one you get from evaluate (or accuracy_score from sklearn). For example, I tried with your configuration and made the corresponding Keras model directly instead of using Talos, and the last epoch result is roughly 2% different from what both evaluate and accuracy_score gives. 

As I now know what you are looking for and your data, I can confirm that this is not an issue in Talos, and even though it is something that is known about Keras (there are many issues discussing this), I'm not sure it can be considered an issue because the last epoch training result and evaluation are really not the same thing. **You get exactly the same level of reproducibility with Talos as you get with Keras. There is absolutely no difference.**
